src/backend/distributed/executor/multi_executor.c

```diff
@@ -24,273 +24,440 @@
 	}
-	else
+
+	scanDirection = scanState->customScanState.ss.ps.state->es_direction;
+	Assert(ScanDirectionIsValid(scanDirection));
+
+	if (ScanDirectionIsBackward(scanDirection))
 	{
-		/* drop into the standard executor */
-		standard_ExecutorEnd(queryDesc);
+		forwardScanDirection = false;
 	}
 
-	/*
-	 * Final step of a distributed query is executing the master node select
-	 * query. We clean up the temp tables after executing it, if we already created it.
-	 */
-	if (eflags & EXEC_FLAG_CITUS_MASTER_SELECT)
+	resultSlot = scanState->customScanState.ss.ps.ps_ResultTupleSlot;
+	tuplestore_gettupleslot(tupleStore, forwardScanDirection, false, resultSlot);
+
+	return resultSlot;
+}
+
+
+/*
+ * TaskTrackerExecScan is a callback function which returns next tuple from a
+ * task-tracker execution. In the first call, it executes distributed task-tracker
+ * plan and loads results from temporary files into custom scan's tuple store.
+ * Then, it returns tuples one by one from this tuple store.
+ */
+TupleTableSlot *
+TaskTrackerExecScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
+	TupleTableSlot *resultSlot = NULL;
+
+	if (!scanState->finishedRemoteScan)
 	{
-		PlannedStmt *planStatement = queryDesc->plannedstmt;
-		int savedLogMinMessages = 0;
-		int savedClientMinMessages = 0;
+		MultiPlan *multiPlan = scanState->multiPlan;
+		Job *workerJob = multiPlan->workerJob;
+
+		PrepareMasterJobDirectory(workerJob);
+		MultiTaskTrackerExecute(workerJob);
 
-		RangeTblEntry *rangeTableEntry = linitial(planStatement->rtable);
-		Oid masterTableRelid = rangeTableEntry->relid;
+		LoadTuplesIntoTupleStore(scanState, workerJob);
 
-		ObjectAddress masterTableObject = { InvalidOid, InvalidOid, 0 };
+		scanState->finishedRemoteScan = true;
+	}
 
-		masterTableObject.classId = RelationRelationId;
-		masterTableObject.objectId = masterTableRelid;
-		masterTableObject.objectSubId = 0;
+	resultSlot = ReadNextTuple(scanState);
 
-		/*
-		 * Temporarily change logging level to avoid DEBUG2 logging output by
-		 * performDeletion. This avoids breaking the regression tests which
-		 * use DEBUG2 logging.
-		 */
-		savedLogMinMessages = log_min_messages;
-		savedClientMinMessages = client_min_messages;
+	return resultSlot;
+}
 
-		log_min_messages = INFO;
-		client_min_messages = INFO;
 
-		performDeletion(&masterTableObject, DROP_RESTRICT, PERFORM_DELETION_INTERNAL);
+/*
+ * CitusEndScan is used to clean up tuple store of the given custom scan state.
+ */
+void
+CitusEndScan(CustomScanState *node)
+{
+	CitusScanState *scanState = (CitusScanState *) node;
 
-		log_min_messages = savedLogMinMessages;
-		client_min_messages = savedClientMinMessages;
+	if (scanState->tuplestorestate)
+	{
+		tuplestore_end(scanState->tuplestorestate);
+		scanState->tuplestorestate = NULL;
 	}
 }
+
+
+/*
+ * CitusReScan is just a place holder for rescan callback. Currently, we don't
+ * support rescan given that there is not any way to reach this code path.
```

Comment:
Whenever we implement functionality like pull to master for CTEs, there would be cases we would like to join two custom scans on the master node. This would lead to a nested loop join which requires rescanning on one of the custom scans.